apps/api-extractor/src/collector/Collector.ts

```diff
@@ -0,0 +1,551 @@
+   *
+   * ```
+   * /// <reference lib="runtime-library" />
+   * ```
+   */
+  public get dtsLibReferenceDirectives(): ReadonlySet<string> {
+    return this._dtsLibReferenceDirectives;
+  }
+
+  public get entities(): ReadonlyArray<CollectorEntity> {
+    return this._entities;
+  }
+
+  /**
+   * Perform the analysis.
+   */
+  public analyze(): void {
+    if (this._astEntryPoint) {
+      throw new Error('DtsRollupGenerator.analyze() was already called');
+    }
+
+    // This runs a full type analysis, and then augments the Abstract Syntax Tree (i.e. declarations)
+    // with semantic information (i.e. symbols).  The "diagnostics" are a subset of the everyday
+    // compile errors that would result from a full compilation.
+    for (const diagnostic of this._program.getSemanticDiagnostics()) {
+      const errorText: string = TypeScriptMessageFormatter.format(diagnostic.messageText);
+      this.reportError(`TypeScript: ${errorText}`, diagnostic.file, diagnostic.start);
+    }
+
+    // Build the entry point
+    const astEntryPoint: AstEntryPoint = this.astSymbolTable.fetchEntryPoint(this.package.entryPointSourceFile);
+
+    const packageDocCommentTextRange: ts.TextRange | undefined = PackageDocComment.tryFindInSourceFile(
+      this.package.entryPointSourceFile, this);
+
+    if (packageDocCommentTextRange) {
+      const range: tsdoc.TextRange = tsdoc.TextRange.fromStringRange(this.package.entryPointSourceFile.text,
+        packageDocCommentTextRange.pos, packageDocCommentTextRange.end);
+
+      this.package.tsdocParserContext = this._tsdocParser.parseRange(range);
+      this.package.tsdocComment = this.package.tsdocParserContext!.docComment;
+    }
+
+    const exportedAstSymbols: AstSymbol[] = [];
+
+    // Create a CollectorEntity for each top-level export
+    for (const exportedMember of astEntryPoint.exportedMembers) {
+      const astSymbol: AstSymbol = exportedMember.astSymbol;
+
+      this._createEntityForSymbol(exportedMember.astSymbol, exportedMember.name);
+
+      exportedAstSymbols.push(astSymbol);
+    }
+
+    // Create a CollectorEntity for each indirectly referenced export.
+    // Note that we do this *after* the above loop, so that references to exported AstSymbols
+    // are encountered first as exports.
+    const alreadySeenAstSymbols: Set<AstSymbol> = new Set<AstSymbol>();
+    for (const exportedAstSymbol of exportedAstSymbols) {
+      this._createEntityForIndirectReferences(exportedAstSymbol, alreadySeenAstSymbols);
+
+      this.fetchMetadata(exportedAstSymbol);
+    }
+
+    this._makeUniqueNames();
+
+    Sort.sortBy(this._entities, x => x.getSortKey());
+    Sort.sortSet(this._dtsTypeReferenceDirectives);
+    Sort.sortSet(this._dtsLibReferenceDirectives);
+  }
+
+  public tryGetEntityBySymbol(symbol: ts.Symbol): CollectorEntity | undefined {
+    return this._entitiesBySymbol.get(symbol);
+  }
+
+  public fetchMetadata(astSymbol: AstSymbol): SymbolMetadata;
```

Comment:
>fetchMetadata [](start = 9, length = 13)

The word `fetch` makes these sound like they're querying the filesystem or the network. `get` is a better verb.

Reply:
I use "get" for lightweight operations, and "fetch" for operations that do nontrivial work to build something if it's not cached yet.  This distinction seems important and useful.  @iclanton can you think of any other verbs?

---
In reply to: [236472082](https://github.com/Microsoft/web-build-tools/pull/952#discussion_r236472082) [](ancestors = 236472082)